File: backend/src/dto/oauth.rs
use gitdot_api::resource::oauth as api;
use gitdot_core::dto::{DeviceCodeResponse, TokenResponse};

use super::IntoApi;

impl IntoApi for DeviceCodeResponse {
    type ApiType = api::DeviceCodeResource;
    fn into_api(self) -> Self::ApiType {
        api::DeviceCodeResource {
            device_code: self.device_code,
            user_code: self.user_code,
            verification_uri: self.verification_uri,
            expires_in: self.expires_in,
            interval: self.interval,
        }
    }
}

impl IntoApi for TokenResponse {
    type ApiType = api::TokenResource;
    fn into_api(self) -> Self::ApiType {
        api::TokenResource {
            access_token: self.access_token,
            user_name: self.user_name,
            user_email: self.user_email,
        }
    }
}

File: frontend/app/lib/dto/oauth.ts
import { z } from "zod";

export const GetDeviceCodeQuerySchema = z.object({
  client_id: z.string(),
});

export type GetDeviceCodeQuery = z.infer<typeof GetDeviceCodeQuerySchema>;

export const DeviceCodeResponseSchema = z.object({
  device_code: z.string(),
  user_code: z.string(),
  verification_uri: z.string(),
  expires_in: z.number(),
  interval: z.number(),
});

export type DeviceCodeResponse = z.infer<typeof DeviceCodeResponseSchema>;

export const PollTokenRequestSchema = z.object({
  device_code: z.string(),
  client_id: z.string(),
});

export type PollTokenRequest = z.infer<typeof PollTokenRequestSchema>;

export const TokenResponseSchema = z.object({
  access_token: z.string(),
  user_name: z.string(),
  user_email: z.string(),
});

export type TokenResponse = z.infer<typeof TokenResponseSchema>;

export const AuthorizeDeviceRequestSchema = z.object({
  user_code: z.string(),
});

export type AuthorizeDeviceRequest = z.infer<
  typeof AuthorizeDeviceRequestSchema
>;
